doc/subsystems/codegen/principle.rst

```diff
@@ -0,0 +1,48 @@
+..
+    Copyright (c) 2018 Bobby Noelte
+    SPDX-License-Identifier: Apache-2.0
+
+.. _codegen_principle:
+
+Code Generation Principle
+#########################
+
+How code generation works in Zephyr.
+
+.. contents::
+   :depth: 2
+   :local:
+   :backlinks: top
+
+Principle
+---------
+
+.. image:: codegen_principle.png
```

Comment:
Seems like these four pictures could use some words to explain what they're showing...